bin/wasm-node/rust/src/json_rpc_service.rs

```diff
@@ -57,72 +57,84 @@ pub async fn request_handling_task(
-                                    "A JSON-RPC service has not been started for chain index {}",
-                                    chain_index
-                                ),
-                                ),
-                                None,
-                            ),
-                            chain_index,
-                        );
-                    }
+                        let (request_id, call) = match methods::parse_json_call(request_str) {
+                            Ok(rq) => rq,
+                            Err(methods::ParseError::Method { request_id, error }) => {
+                                log::warn!(
+                                    target: "json-rpc",
+                                    "Error in JSON-RPC method call: {}", error
+                                );
+                                send_back(&error.to_json_error(request_id), chain_index, source_id);
+                                return;
+                            }
+                            Err(error) => {
+                                log::warn!(
+                                    target: "json-rpc",
+                                    "Ignoring malformed JSON-RPC call: {}", error
+                                );
+                                return;
+                            }
+                        };
+
+                        match json_rpc_services.get(&chain_index).cloned() {
+                            Some(service) => service.handle_rpc(source_id, request_id, call).await,
+                            None => {
+                                send_back(
+                                    &json_rpc::parse::build_error_response(
+                                        request_id,
+                                        json_rpc::parse::ErrorResponse::ApplicationDefined(
+                                            -33000,
+                                            &format!(
+                                            "A JSON-RPC service has not been started for chain index {}",
+                                            chain_index
+                                        ),
+                                        ),
+                                        None,
+                                    ),
+                                    chain_index,
+                                    source_id
+                                );
+                            }
+                        }
+                    }));
+                }
+                ffi::JsonRpcMessage::UnsubscribeAll { source_id } => {
+                    let json_rpc_services = json_rpc_services.clone();
+                    (tasks_executor.lock().await)(Box::pin(async move {
+                        for service in json_rpc_services.values().cloned() {
+                            service.handle_unsubscribe_all(source_id).await;
+                        }
+                    }));
```

Comment:
```suggestion
                        for service in json_rpc_services.values().cloned() {
                            service.handle_unsubscribe_all(source_id).await;
                        }
```

You're trying to solve race conditions. We shouldn't spawn move asynchronous tasks!